server/utils.py

```diff
@@ -91,3 +95,117 @@ def mask_string(s: str) -> str:
+
+    # incoming teams list should be same as rostered teams list
+    missing_teams = set(rostered_team_ids) - set(incoming_teams)
+    if missing_teams:
+        errors["missing_teams"] = list(missing_teams)
+        valid_seeds_and_teams = False
+
+    # all teams in the update should belong to the roster
+    wrong_teams = set(incoming_teams) - set(rostered_team_ids)
+    if wrong_teams:
+        valid_seeds_and_teams = False
+        errors["wrong_teams"] = list(wrong_teams)
+
+    # same team cannot be in 2 different seeds
+    duplicate_teams: list[int] = [
+        team for team, occurence in Counter(incoming_teams).items() if occurence > 1
+    ]
+    if duplicate_teams:
+        valid_seeds_and_teams = False
+        errors["duplicate_teams"] = duplicate_teams
+
+    return valid_seeds_and_teams, errors
+
+
+def validate_new_pool(
+    tournament: Tournament, new_pool: set[int]
+) -> tuple[bool, validation_error_dict]:
+    # same seed shouldn't be added in multiple pools
+    other_pools_in_tournament = Pool.objects.filter(tournament=tournament).values_list(
+        "initial_seeding", flat=True
+    )
+    already_present_seeds: set[int] = set()
+    for pool in other_pools_in_tournament:
+        already_present_seeds.update(
+            map(int, pool.keys())
+        )  # Pool.initial_seeding : dict[seed(str), team_id(int)]. Convert seeds from str to ints
+
+    repeated_seeds_in_new_pool = already_present_seeds.intersection(new_pool)
+    # the seed shouldn't negative, 0 or more than roster size
+    tournament_roster_size = (
+        UCRegistration.objects.filter(event=tournament.event)
+        .values_list("team", flat=True)
+        .distinct()
+        .count()
+    )
+    invalid_seeds = list(filter(lambda seed: 1 <= seed <= tournament_roster_size, new_pool))
+    valid_pool = True
+    errors: validation_error_dict = {}
+
+    if repeated_seeds_in_new_pool:
+        errors["repeated_seeds_in_new_pool"] = list(repeated_seeds_in_new_pool)
+        valid_pool = False
+    if invalid_seeds:
+        errors["invalid_seeds"] = invalid_seeds
+        valid_pool = False
+
+    return valid_pool, errors
+
+
+POSITION_MATCH_BRACKET_SIZE = 2
+SEMIFINAL_BRACKET_SIZE = 4
+QUARTERFINAL_BRACKET_SIZE = 8
+THIRD_PLACE_MATCH_SEED = 3
+
+
+def get_bracket_match_name(start: int, end: int, seed_1: int, seed_2: int) -> str:
+    """Get type of bracket (Quaterfinals, Semifinals, Finals, Position match)"""
+    bracket_size = (end - start) + 1
+    seed_1 = min(seed_1, seed_2)
+    seed_2 = max(seed_1, seed_2)
+
+    if bracket_size == POSITION_MATCH_BRACKET_SIZE:
+        if seed_1 == 1:
+            return "Finals"
+        if seed_1 == THIRD_PLACE_MATCH_SEED:
+            return "3rd Place Match"
+
+    if bracket_size == SEMIFINAL_BRACKET_SIZE and 1 <= seed_1 <= bracket_size:
+        return f"SF {seed_1} v {seed_2}"
+
+    if bracket_size == QUARTERFINAL_BRACKET_SIZE and 1 <= seed_1 <= bracket_size:
+        return f"QF {seed_1} v {seed_2}"
```

Comment:
Should we just have the name as "Quarter Finals" here as we anyway can get the seeds from placeholder fields?

Reply:
Yeah, same as above answer, we can.

Reply:
so this function can return full name with seed right ? no need to change ?

Reply:
[same like here](https://github.com/india-ultimate/hub/pull/145#discussion_r1370030016)